README.org
* org-wiki-tools

Help maintaining order & beauty in your org-mode-wiki

** Usage

#+BEGIN_SRC bash
  $ vim config.edn
  $ ./extract-links.sh
  $ ./process-links.sh
  $ ./proocess-wiki.sh
#+END_SRC

** Test

#+BEGIN_SRC bash
  $ lein test
#+END_SRC


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/org-wiki-tools/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/org-wiki-tools/compare/0.1.0...0.1.1


test/org_wiki_tools/core_test.clj
(ns org-wiki-tools.core-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :as io]
            [clj-time.local :as l]
            [clj-http.client :as client]
            [me.raynes.fs :as fs]
            [org-wiki-tools.core :refer :all]))

;; vars
(def normal-line "* [[https://example.com][Test]] :archive:")
(def no-link-line "* No link folk!")
(def wikipedia-link "https://en.wikipedia.org/wiki/Sweet_corn")
(def wikiwand-link "https://www.wikiwand.com/en/Sweet_corn")

;; helpers
(defn make-map [length]
  (into {} (map #(hash-map %1 %2) (range length) (range length))))

(testing "info enabled"
  (with-redefs [println str
                print-info true]
    (= (info "test" "test") "testtest")))

(testing "info disabled"
  (with-redefs [println str
                print-info false]
    (= (info "test" "test") nil)))

;; list-org-files
(testing "list-org-files"
  (with-redefs [fs/glob (fn [_] (vector "test.org" "bar.org" "foo.org"))]
    (= (list-org-files "./org") ["test.org" "bar.org" "foo.org"])))

;; get-link-info
(testing "get-link-info 200"
  (with-redefs [client/get (fn [_ _] (hash-map :status 200))]
    (= {:status 200 :url "https://example.com"} (get-link-info {:url "https://example.com"}))))
;; TODO get-link-info 404

;; process-link
(testing "process-link"
  (with-redefs [l/local-now #(str "1970-01-01")
                get-link-info (fn [_] (hash-map :url "https://example.com"))]
    (= (process-link {:url "https://example.com"}) {"https://example.com" {:url "https://example.com" :processed-at "1970-01-01"}})))

(testing "tag-dead-link dead"
  (with-redefs [link-db {"https://example.com" {:status 404}}]
    (= (str "* [[https://example.com][Example]] :archive:" "\t\t\t" ":dead-link:") (tag-dead-link "* [[https://example.com][Example]] :archive:"))))

(testing "tag-dead-link alive"
  (with-redefs [link-db {"https://example.com" {:status 200}}]
    (= (str "* [[https://example.com][Example]] :archive:") (tag-dead-link "* [[https://example.com][Example]] :archive:"))))

(testing "tag-dead-link unfound"
  (with-redefs [link-db {"https://test.com" {:status 404}}]
    (= (str "* [[https://example.com][Example]] :archive:") (tag-dead-link "* [[https://example.com][Example]] :archive:"))))

(testing "tag-dead-link already dead"
  (with-redefs [link-db {"https://example.com" {:status 404}}]
    (= (str "* [[https://example.com][Example]] :dead-link:") (tag-dead-link "* [[https://example.com][Example]] :dead-link:"))))

;; link->map
(testing "link->map Normal link"
  (is (= {"https://example.com" {:url "https://example.com" :status nil :processed-at nil :title "Test" :site "example.com"}} (link->map normal-line))))
(testing "link->map No link"
  (is (not (link->map no-link-line))))

;; find-link
(testing "find-link Normal link"
  (is (= ["[[https://example.com][Test]]" "https://example.com" "Test"] (find-link normal-line))))
(testing "find-link No link"
  (is (= nil (find-link no-link-line))))

;; wikiwand->wikipedia
(testing "wikiwand->wikipedia"
  (is (= (wikiwand->wikipedia wikiwand-link) wikipedia-link))
  (is (= "nothing" (wikiwand->wikipedia "nothing"))))

;; select-subset
(testing "select-subset"
  (is (= (count (select-subset (make-map (* 29 10)))) 10)))

;; ->archive
(testing "->archive"
  (is (= "* whatever :archived:" (->archive "* whatever :archive:")))
  (is (= "* no nothing" (->archive "* no nothing"))))


src/org_wiki_tools/createdb.clj
(ns org-wiki-tools.createdb
  (:require [datomic.api :as d]))

(def schema
  [{:db/doc "Link url"
    :db/ident :link/url
    :db/valueType :db.type/string
    :db/cardinality :db.cardinality/one
    :db/unique :db.unique/identity
    :db.install/_attribute :db.part/db}

   {:db/doc "Link site"
    :db/ident :link/site
    :db/valueType :db.type/string
    :db/cardinality :db.cardinality/one
    :db.install/_attribute :db.part/db}

   {:db/doc "Link status"
    :db/ident :link/status
    :db/valueType :db.type/long
    :db/cardinality :db.cardinality/one
    :db.install/_attribute :db.part/db}

   {:db/doc "Link body"
    :db/ident :link/body
    :db/valueType :db.type/string
    :db/cardinality :db.cardinality/one
    :db.install/_attribute :db.part/db}

   {:db/doc "Link title"
    :db/ident :link/title
    :db/valueType :db.type/string
    :db/cardinality :db.cardinality/one
    :db.install/_attribute :db.part/db}])

(defn create-schema [db-uri]
  (-> db-uri
      d/connect
      (d/transact schema)))

(defn create-db []
  (let [db-uri "datomic:dev://localhost:4334/test-org?password=help"]
    (println "Deleting DB...")
    ;;(d/delete-database db-uri)
    (println "Creating DB...")
    (d/create-database db-uri)
    (System/exit 0)
    (println "Creating Schema...")
    (create-schema db-uri)
    (println "Done!")
    (System/exit 0)))

(defn -main [& args]
  (create-db))


src/org_wiki_tools/core.clj
(ns org-wiki-tools.core
  (:require [clojure.edn :as edn]
            [clj-time.local :as l]
            [clojure.java.io :as io]
            [clj-http.client :as client]
            [com.climate.claypoole :as cp]
            [me.raynes.fs :as fs]))

(def link-db-path "./link.edn")
(def link-db {})

;; Load config and link-db
(def config (-> "./config.edn" slurp clojure.edn/read-string))
(try
  (def link-db (-> "./link.edn" slurp clojure.edn/read-string))
  (catch Exception ex
    (def link-db {})
    (println ex)))

;; Regex
(def org-mode-link-re #"\[?\[?(http[^]]*)\]?\[?([^]]*)\]?\]?")
(def site-re #"https?:\/\/([^/]*)/?")

;; Globals
(def print-info (:print-info config true))
(def link-processing-limit (:link-processing-limit config 29))
(def wiki-path (:wiki-path config))
(def link-to-archive (atom []))
(def new-link-db (atom {}))

;; REPL
;; (count link-db)
;; (count (filter #(nil? (:processed-at %)) (vals link-db)))
;; (println (filter #(= 404 (:status %)) (vals link-db)))

(defn info [& msg]
  (when print-info
    (apply println msg)))

(defn list-org-files [dir]
  (fs/glob (str dir "/*.org")))

(defn get-link-info [link]
  (merge link
         (try
           (select-keys (client/get (:url link) {:socket-timeout 10000 :connection-timeout 10000}) [:status])
           (catch Exception ex
             (info "caught exception: " (.getMessage ex))
             {:status (get-in (Throwable->map ex) [:data :status] 404) :error (.getMessage ex)}))))

(defn link->map [line]
  (when-let [[_ url title] (re-find org-mode-link-re line)]
    (let [[_ site] (re-find site-re url)]
      (info (str "Link: " url))
      {(str url) {:url url :status nil :processed-at nil :title title :site site}})))

(defn process-link [link]
  (info "Processing Link: " link)
  {(:url link) (merge (get-link-info link) {:processed-at (str (l/local-now))})})

(defn map-file [fn file]
  (info "Extracting link from file: " file)
  (with-open [rdr (io/reader file)]
    (doall (map fn (line-seq rdr)))))

(defn get-links [file]
  (info "Extracting link from file: " file)
  (with-open [rdr (io/reader file)]
    (into {} (map link->map (line-seq rdr)))))

(defn find-link [text]
  (re-find org-mode-link-re text))

(defn wikiwand->wikipedia [link]
  (clojure.string/replace link #"https:\/\/www\.wikiwand\.com\/(.{2})\/(.*)" (fn [[_ l k]] (format "https://%s.wikipedia.org/wiki/%s" l k))))

(defn tag-dead-link [line]
  (if-let [[_ url _] (find-link line)]
    (if (and (not= 200 (:status (link-db url) 200)) (not (re-find #":dead-link:" line)))
      (str line "\t\t\t" ":dead-link:")
      line)
    line))

(defn ->archive [line]
  (if (re-find #":archive:" line)
    (do
      (swap! link-to-archive conj (second (find-link line)))
      (clojure.string/replace line #":archive:" ":archived:"))
    line))

(defn select-subset [links]
  (let [nb (int (/ (count links) link-processing-limit))]
    (take nb (filter #(nil? (:processed-at %)) links))))

(defn extract-links [dir]
  (let [files (list-org-files dir)]
    (reset! new-link-db (into {} (map get-links files)))))

(defn process-wiki [dir]
  (let [files (list-org-files dir)]
    (doall
     (map
      #(->> %
            (map-file wikiwand->wikipedia)
            (map ->archive)
            (map tag-dead-link)
            (clojure.string/join "\n")
            (spit %))
      files)))
  (spit "to_archive.txt" (str (clojure.string/join "\n" @link-to-archive) "\n")))

(defn process-links []
  (->> link-db
       vals
       select-subset
       doall
       (cp/pmap 100 process-link)
       (into {})
       (merge link-db)
       prn-str
       (spit link-db-path)))

;; TODO Autoheal link (redirect)
;; TODO Crontab ./extract-link.sh monthly
;; TODO Crontab ./process-link.sh daily
;; TODO Crontab ./process-wiki.sh daily

(defn -main [& args]
  (condp = (first args)
    "show-link" (info link-db)
    "process-links" (do
                     (info "Processing link...")
                     (process-links))
    "process-wiki" (do
                     (info "Processing wiki...")
                     (process-wiki wiki-path))
    "extract-links" (do
                (info "Extracting...")
                (extract-links wiki-path)
                (spit link-db-path (prn-str @new-link-db))))
  (println "Done!")
  (System/exit 0))
